File: repair_management_platform/repair_management_platform/report/repair_analytics/repair_analytics.js
frappe.query_reports["Repair Analytics"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -3),
            "reqd": 0
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 0
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nDraft\nDiagnostic Pending\nDiagnostic Completed\nQuoted\nRepair in Progress\nReady for Collection\nCompleted\nCancelled"
        },
        {
            "fieldname": "industry",
            "label": __("Industry"),
            "fieldtype": "Link",
            "options": "Industry Type"
        },
        {
            "fieldname": "technician",
            "label": __("Technician"),
            "fieldtype": "Link",
            "options": "Employee"
        },
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer"
        },
        {
            "fieldname": "priority",
            "label": __("Priority"),
            "fieldtype": "Select",
            "options": "\nLow\nMedium\nHigh\nUrgent"
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname == "status") {
            if (value == "Completed") {
                value = "<span class='indicator-pill green'>" + value + "</span>";
            } else if (value == "Cancelled") {
                value = "<span class='indicator-pill red'>" + value + "</span>";
            } else if (value == "Repair in Progress") {
                value = "<span class='indicator-pill blue'>" + value + "</span>";
            } else if (value == "Ready for Collection") {
                value = "<span class='indicator-pill orange'>" + value + "</span>";
            }
        }

        if (column.fieldname == "priority") {
            if (value == "Urgent") {
                value = "<span class='indicator-pill red'>" + value + "</span>";
            } else if (value == "High") {
                value = "<span class='indicator-pill orange'>" + value + "</span>";
            } else if (value == "Medium") {
                value = "<span class='indicator-pill yellow'>" + value + "</span>";
            } else if (value == "Low") {
                value = "<span class='indicator-pill grey'>" + value + "</span>";
            }
        }

        return value;
    }
};

File: repair_management_platform/repair_management_platform/report/repair_analytics/repair_analytics.py
import frappe
from frappe import _
from frappe.utils import flt, getdate

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    return [
        {
            "fieldname": "repair_order",
            "label": _("Repair Order"),
            "fieldtype": "Link",
            "options": "Repair Order",
            "width": 150
        },
        {
            "fieldname": "creation_date",
            "label": _("Creation Date"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "customer",
            "label": _("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "width": 150
        },
        {
            "fieldname": "device",
            "label": _("Device"),
            "fieldtype": "Link",
            "options": "Repair Device",
            "width": 150
        },
        {
            "fieldname": "industry",
            "label": _("Industry"),
            "fieldtype": "Link",
            "options": "Industry Type",
            "width": 120
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "priority",
            "label": _("Priority"),
            "fieldtype": "Data",
            "width": 80
        },
        {
            "fieldname": "technician",
            "label": _("Technician"),
            "fieldtype": "Link",
            "options": "Employee",
            "width": 150
        },
        {
            "fieldname": "total_amount",
            "label": _("Total Amount"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "paid_amount",
            "label": _("Paid Amount"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "balance",
            "label": _("Balance"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "turnaround_time_days",
            "label": _("Turnaround (Days)"),
            "fieldtype": "Int",
            "width": 100
        },
        {
            "fieldname": "estimated_completion_date",
            "label": _("Est. Completion"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "actual_completion_date",
            "label": _("Actual Completion"),
            "fieldtype": "Date",
            "width": 100
        }
    ]

def get_data(filters):
    conditions = get_conditions(filters)

    data = frappe.db.sql("""
        SELECT
            ro.name as repair_order,
            DATE(ro.creation) as creation_date,
            ro.customer,
            ro.device,
            ro.industry,
            ro.status,
            ro.priority,
            ro.technician,
            ro.total_amount,
            ro.paid_amount,
            (ro.total_amount - IFNULL(ro.paid_amount, 0)) as balance,
            ro.turnaround_time_days,
            ro.estimated_completion_date,
            ro.actual_completion_date
        FROM
            `tabRepair Order` ro
        WHERE
            ro.docstatus < 2
            {conditions}
        ORDER BY
            ro.creation DESC
    """.format(conditions=conditions), filters, as_dict=1)

    return data

def get_conditions(filters):
    conditions = []

    if filters.get("from_date"):
        conditions.append("DATE(ro.creation) >= %(from_date)s")

    if filters.get("to_date"):
        conditions.append("DATE(ro.creation) <= %(to_date)s")

    if filters.get("status"):
        conditions.append("ro.status = %(status)s")

    if filters.get("industry"):
        conditions.append("ro.industry = %(industry)s")

    if filters.get("technician"):
        conditions.append("ro.technician = %(technician)s")

    if filters.get("customer"):
        conditions.append("ro.customer = %(customer)s")

    if filters.get("priority"):
        conditions.append("ro.priority = %(priority)s")

    return " AND " + " AND ".join(conditions) if conditions else ""
